### PR TITLE
Add common functions

### LINKED ISSUE #93: Add "basic URL file" to create_package_template.py
Add "basic URL file" to create_package_template.py and to the issue template and GH action that create PRs from issues.

### PR DESCRIPTION
**Pull request by day1player**
Added "VM-Install-Single-Exe" and "VM-Install-Single-Ps1" functions to common module. Also created templates for each of these types of packages in the python helper script.

closes #93

### CODE REVIEW

File: packages/common.vm/tools/vm.common/vm.common.psm1 [hunk 1]
```diff
@@ -496,6 +496,101 @@ function VM-Install-From-Zip {
...
+    [Parameter(Mandatory=$true, Position=1)]
+    [string] $category,
+    [Parameter(Mandatory=$true, Position=2)]
+    [string] $ps1Url,
+    [Parameter(Mandatory=$false)]
+    [string] $ps1Sha256,
+    [Parameter(Mandatory=$false)]
+    [string] $ps1Url_64,
+    [Parameter(Mandatory=$false)]
+    [string] $ps1Sha256_64
```
Comment (MalwareMechanic) [inline packages/common.vm/tools/vm.common/vm.common.psm1:560]: Allow users to specify an additional PowerShell command to execute after changing into the script directory. For instance: `Import-Module .\WMIOps.ps1; Get-Content README.md`

```suggestion
    [string] $ps1Sha256_64,
    [Parameter(Mandatory=$false)]
    [string] $ps1Cmd
```

File: packages/common.vm/tools/vm.common/vm.common.psm1 [hunk 2]
```diff
@@ -496,6 +496,101 @@ function VM-Install-From-Zip {
...
+    # Create shortcut
+    $target_cmd = Join-Path ${Env:WinDir} "system32\cmd.exe" -Resolve
+    $target_args = '/K powershell.exe -ExecutionPolicy Bypass -NoExit -Command "cd ' + $toolDir + '"'
+    $target_dir = $toolDir
+    $target_icon = Join-Path (Join-Path ${Env:WinDir} "system32\WindowsPowerShell\v1.0") "powershell.exe" -Resolve
+
+    $shortcut = Join-Path $shortcutDir ($toolName + ".ps1.lnk")
+    Install-ChocolateyShortcut -shortcutFilePath $shortcut -targetPath $target_cmd -Arguments $target_args -WorkingDirectory $target_dir -IconLocation $target_icon -RunAsAdmin
+    VM-Assert-Path $shortcut
+
```
Comment (MalwareMechanic) [inline packages/common.vm/tools/vm.common/vm.common.psm1:588]: Allow users to specify an additional PowerShell command to execute after changing into the script directory. For instance: `Import-Module .\WMIOps.ps1; Get-Content README.md`. Used the `$ps1Cmd` here to allow this.

I also tried to consider paths that have spaces in them and properly quote -- initial testing seems good, but we should test later.

```suggestion
    $shortcut = Join-Path $shortcutDir "$toolName.ps1.lnk"
    $targetCmd = Join-Path ${Env:WinDir} "system32\cmd.exe" -Resolve
    if ($ps1Cmd) {
      $targetArgs = "/K powershell.exe -ExecutionPolicy Bypass -NoExit -Command `"cd '$toolDir'; $ps1Cmd`""
    } else {
      $targetArgs = "/K powershell.exe -ExecutionPolicy Bypass -NoExit -Command `"cd '$toolDir'`""
    }
    $targetIcon = Join-Path (Join-Path ${Env:WinDir} "system32\WindowsPowerShell\v1.0") "powershell.exe" -Resolve
    
    Install-ChocolateyShortcut -shortcutFilePath $shortcut -targetPath $targetCmd -arguments $targetArgs -workingDirectory $toolDir -iconLocation $targetIcon
    VM-Assert-Path $shortcut

```

**Review by MalwareMechanic — CHANGES_REQUESTED** [review]

File: scripts/utils/create_package_template.py [hunk 3]
```diff
@@ -27,7 +27,7 @@
     "Hex Editors",
     "Java",
     "Javascript",
-    "Net",
+    "Networking Tools",
```
Comment (MalwareMechanic) [inline scripts/utils/create_package_template.py:30]: Do we need the word "Tools"?

```suggestion
    "Networking",
```

File: scripts/utils/create_package_template.py [hunk 4]
```diff
@@ -37,6 +37,16 @@
...
+    "Active Directory Tools",
+    "Command & Control",
+    "Evasion",
+    "Exploitation",
+    "Information Gathering",
+    "Networking Tools",
+    "Password Attacks",
+    "Utilities",
+    "Vulnerability Analysis",
+    "Wordlists"
```
Comment (MalwareMechanic) [inline scripts/utils/create_package_template.py:49]: Removed redundant "Networking Tools". Also do we need the word "Tools"?

```suggestion
    "Active Directory",
    "Command & Control",
    "Evasion",
    "Exploitation",
    "Information Gathering",
    "Password Attacks",
    "Utilities",
    "Vulnerability Analysis",
    "Wordlists"
```

**Review by MalwareMechanic — CHANGES_REQUESTED** [review]

**Review by Ana06 — CHANGES_REQUESTED** [review]
Nice functions additions! 😄 

Please ensure the tests work, this may be useful: https://github.com/mandiant/VM-Packages/wiki/Contributing#testing

**Comment by Ana06** [comment]
What's the state of this PR? @day1player do you need help addressing the requested changes?

@MalwareMechanic what is the state of the category names?

**Review by MalwareMechanic — CHANGES_REQUESTED** [review]

File: scripts/utils/create_package_template.py [hunk 5]
```diff
@@ -37,6 +37,15 @@
     "Utilities",
     "VB",
     "Web Application",
+    "Active Directory",
+    "Command & Control",
+    "Evasion",
+    "Exploitation",
+    "Information Gathering",
+    "Password Attacks",
+    "Utilities",
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply scripts/utils/create_package_template.py:46]: Yeah, I left these untouched for now as I have not heard back on #38

File: scripts/utils/create_package_template.py [hunk 6]
```diff
@@ -82,37 +91,53 @@
...
+
+$toolName = '{tool_name}'
+$category = '{category}'
+
+$toolUrl = '{tool_url}'
+$toolSha256 = '{tool_hash}'
+
+VM-Install-Executable $toolName $category $toolUrl $toolSha256
+"""
+
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply scripts/utils/create_package_template.py:140]: Yes, also an artifact of my git messup.

File: packages/common.vm/tools/vm.common/vm.common.psm1 [hunk 7]
```diff
@@ -496,6 +496,217 @@ function VM-Install-From-Zip {
...
+    $shortcut = Join-Path $shortcutDir "$toolName.ps1.lnk"
+    $targetCmd = Join-Path ${Env:WinDir} "system32\cmd.exe" -Resolve
+    if ($ps1Cmd) {
+      $targetArgs = "/K powershell.exe -ExecutionPolicy Bypass -NoExit -Command `"cd '$toolDir'; $ps1Cmd`""
+    } else {
+      $targetArgs = "/K powershell.exe -ExecutionPolicy Bypass -NoExit -Command `"cd '$toolDir'`""
+    }
+    $targetIcon = Join-Path (Join-Path ${Env:WinDir} "system32\WindowsPowerShell\v1.0") "powershell.exe" -Resolve
+
+    Install-ChocolateyShortcut -shortcutFilePath $shortcut -targetPath $targetCmd -arguments $targetArgs -workingDirectory $toolDir -iconLocation $targetIcon
```
[the remark this answers is not in the document]
Comment (mr-tz) [reply packages/common.vm/tools/vm.common/vm.common.psm1:606]: Yes

File: scripts/utils/create_package_template.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mr-tz) [reply scripts/utils/create_package_template.py]: Removed for now, but still waiting on final word on #38.

File: scripts/utils/create_package_template.py [hunk 8]
```diff
@@ -27,7 +27,7 @@
     "Hex Editors",
     "Java",
     "Javascript",
-    "Net",
+    "Networking",
```
Comment (Ana06) [inline scripts/utils/create_package_template.py:30]: are these the final categories? They are different from https://github.com/mandiant/VM-Packages/issues/38#issuecomment-1313874710

As already mentioned in this PR, modifying the categories also needs to update our current GH actions and the packages that are using the old categories (and maybe somewhere else?).

**Review by Ana06 — CHANGES_REQUESTED** [review]
We should squash the commits either manually or using `Squash and merge`.

File: scripts/utils/create_package_template.py (hunk 8 above)
Comment (mr-tz) [reply scripts/utils/create_package_template.py:30]: These are not final, no. Will handle that as part of #38.

File: scripts/utils/create_package_template.py [hunk 9]
```diff
@@ -327,8 +438,8 @@ def main(argv=None):
     parser.add_argument("--category", type=str, choices=CATEGORIES, help="Category for tool")
     parser.add_argument("--description", type=str, default="", help="Description for tool")
     parser.add_argument("--dependency", type=str, default="", help="Metapackage dependency")
-    parser.add_argument("--zip_url", type=str, default="", help="URL to ZIP file")
-    parser.add_argument("--zip_hash", type=str, default="", help="SHA256 hash of ZIP file")
+    parser.add_argument("--target_url", type=str, default="", help="URL to target file (zip or executable)")
+    parser.add_argument("--target_hash", type=str, default="", help="SHA256 hash of target file (zip or executable)")
     parser.add_argument("--shim_path", type=str, default="", help="Metapackage shim path")
```
Comment (mr-tz) [inline scripts/utils/create_package_template.py:443]: Do we need to add `--target_url_64` and `--target_hash_64` here?

File: scripts/utils/create_package_template.py (hunk 8 above)
Comment (mr-tz) [reply scripts/utils/create_package_template.py:27]: Updated them now per the last comment in #38.

File: scripts/utils/create_package_template.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mr-tz) [reply scripts/utils/create_package_template.py]: Updated these now.

File: scripts/utils/create_package_template.py (hunk 9 above)
Comment (MalwareMechanic) [reply scripts/utils/create_package_template.py:452]: Based on https://docs.chocolatey.org/en-us/create/functions/get-chocolateywebfile#url-string it seems we're *supposed* to use the correct URL based on the bitness of the tool we're downloading. But in practice, VMs will likely be 64-bit and I believe the default `-url` parameter should be fine. So we could technically remove the `x64` counterparts from the code to download in the template. Thoughts?

File: scripts/utils/create_package_template.py [hunk 10]
```diff
@@ -277,14 +343,68 @@ def get_script_directory():
...
+            "tool_name",
+            "category",
+            "target_url",
+            "target_hash",
+        ],
     },
     "METAPACKAGE": {
         "cb": create_metapackage_template,
+        "doc": "Install and configure existing packages via dependencies",
+        "example": "install a <tool> already available via chocolatey.org",
```
Comment (MalwareMechanic) [inline scripts/utils/create_package_template.py:407]:
```suggestion
        "example": "Install a <tool> already available via chocolatey.org",
```

File: scripts/utils/create_package_template.py [hunk 11]
```diff
@@ -277,14 +343,68 @@ def get_script_directory():
...
+            "category",
+            "target_url",
+            "target_hash",
+        ],
     },
     "GITHUB_REPO": {
         "cb": create_git_repo_template,
-        "arguments": ["pkg_name", "version", "authors", "description", "tool_name", "category", "zip_url", "zip_hash"],
+        "doc": "A script or code distributed via GitHub in a ZIP file",
+        "example": "https://github.com/.../tool.zip",
```
Comment (MalwareMechanic) [inline scripts/utils/create_package_template.py:392]:
```suggestion
        "example": "https://github.com/sense-of-security/ADRecon/archive/38e4abae3e26d0fa87281c1d0c65cabd4d3c6ebd.zip",
```

File: scripts/utils/create_package_template.py [hunk 12]
```diff
@@ -277,14 +343,68 @@ def get_script_directory():
...
+            "tool_name",
+            "category",
+            "target_url",
+            "target_hash",
+        ],
     },
     "GITHUB_REPO": {
         "cb": create_git_repo_template,
-        "arguments": ["pkg_name", "version", "authors", "description", "tool_name", "category", "zip_url", "zip_hash"],
+        "doc": "A script or code distributed via GitHub in a ZIP file",
```
Comment (MalwareMechanic) [inline scripts/utils/create_package_template.py:391]:
```suggestion
        "doc": "Download a GitHub repository based on a specific commit hash.",
```

**Review by MalwareMechanic — CHANGES_REQUESTED** [review]

File: scripts/utils/create_package_template.py (hunk 9 above)
Comment (mr-tz) [reply scripts/utils/create_package_template.py:452]: I'll leave them there for now, but won't add the `_64` arguments here.

**Review by MalwareMechanic — APPROVED** [review]

**Pull request merged**